src/Stream.elm

```diff
@@ -203,40 +203,28 @@ sample f varying events =
           (Native.Signal.initialValue varying, fromVarying varying)
 
       sampleEvents =
-          Native.Signal.genericMerge (\(Sample s) (Update u) -> SampleAndUpdate s u)
-            (map Sample events)
-            (map Update varyingUpdates)
+          Native.Signal.genericMerge (\s u -> (fst s, snd u))
+            (map (\s -> (Just s, Nothing)) events)
+            (map (\u -> (Nothing, Just u)) varyingUpdates)
   in
       fold sampleUpdate { state = initialValue, trigger = Nothing } sampleEvents
         |> fromVarying
         |> filterMap (\state -> Maybe.map (f state.state) state.trigger)
 
 
-type SampleEvent a b = Sample a | Update b | SampleAndUpdate a b
+type alias SampleEvent b a = (Maybe b, Maybe a)
```

Comment:
I think this is far enough from convention to be pretty confusing. I'd either change the order of things such that it just matches (switching the args in `genericMerge`) or not bother with trying to get `a` and `b` to match.